File: flatten_multilevel_linked_list.py
from __future__ import annotations
from collections import deque

class Node:
    def __init__(self, val: int, prev: Node | None = None, next: Node | None = None, child: Node | None = None):
        self._val = val
        self._prev = prev
        self._next = next
        self._child = child

    @property
    def val(self) -> int:
        return self._val

    @val.setter
    def val(self, val: int):
        if not isinstance(val, (int,)):
            raise TypeError("val must be an int")
        self._val = val

    @property
    def prev(self) -> Node | None:
        return self._prev

    @prev.setter
    def prev(self, value: Node | None):
        if not isinstance(value, (Node, type(None))):
            raise TypeError("prev must be a Node or None")
        self._prev = value

    @property
    def next(self) -> Node | None:
        return self._next

    @next.setter
    def next(self, value: Node | None):
        if not isinstance(value, (Node, type(None))):
            raise TypeError("next must be a Node or None")
        self._next = value

    @property
    def child(self) -> Node | None:
        return self._child

    @child.setter
    def child(self, value: Node | None):
        if not isinstance(value, (Node, type(None))):
            raise TypeError("child must be a Node or None")
        self._child = value


def flatten_multilevel_linked_list_recursive(head: Node | None) -> Node:
    if not head:
        return None
    runner = head
    while runner is not None:
        if runner.child is not None:
            flattened = flatten_multilevel_linked_list_recursive(runner.child)

            temp = runner.next
            
            runner.next = flattened
            flattened.prev = runner
            
            while flattened.next is not None:
                flattened = flattened.next

            if temp:
                flattened.next = temp
                temp.prev = flattened

            runner.child = None

        runner = runner.next
    
    return head 

def flatten_multilevel_linked_list_iterative(head: Node | None) -> Node:
    if not head:
        return None
    stack_ = deque()
    runner = head
    while runner is not None:
        if runner.child is not None:
            child_ = runner.child

            if runner.next is not None:
                stack_.append(runner.next)
            
            runner.next = child_
            child_.prev = runner
            
        if (runner.next is None) and len(stack_) > 0:
            temp = stack_.pop()
            runner.next = temp
            temp.prev = runner
        
        runner = runner.next
    
    return head     

def print_linked_list(head: Node| None):
    runner = head
    sofar = ""
    while runner is not None:
        sofar += f"{runner.val} -> "
        runner = runner.next
    sofar += "None"
    print(sofar)

def main():
    node1 = Node(val=1)
    node2 = Node(val=2)
    node3 = Node(val=3)
    node4 = Node(val=4)
    node5 = Node(val=5)
    node6 = Node(val=6)
    node7 = Node(val=7)
    node8 = Node(val=8)
    node9 = Node(val=9)
    node10 = Node(val=10)
    node11 = Node(val=11)
    node12 = Node(val=12)
    node13 = Node(val=13)
    
    node1.next = node2
    node2.next = node3
    node3.next = node4
    node4.next = node5
    node5.next = node6
    node7.next = node8
    node8.next = node9
    node9.next = node10
    node11.next = node12

    node3.child = node7
    node8.child = node11

    node12.child = node13

    head = flatten_multilevel_linked_list_iterative(node1)
    print_linked_list(head)


if __name__ == '__main__':
    main()
    

File: flatten_multilevel_linkedlist.rs
use std::cell::RefCell;
use std::rc::Rc;

type NodeRef = Option<Rc<RefCell<Node>>>;

struct Node {
    val: i32,
    prev: NodeRef,
    next: NodeRef,
    child: NodeRef,
}

impl Node {
    fn new(val: i32) -> NodeRef {
        Some(Rc::new(RefCell::new(
            Node { val, prev: None, next: None, child: None }
        )))
    } 
}

fn flatten(head: NodeRef) -> NodeRef {
    let mut runner = head.clone();
    let mut stack = Vec::new();

    while let Some(current) = runner.clone() {
        let mut current_borrow = current.borrow_mut();

        // if child exists, flatten it
        if let Some(child) = current_borrow.child.take() {
            // push the next node to the stack, to process it later once the it is done processing the child
            if let Some(next) = current_borrow.next.take() {
                stack.push(next);
            }

            // create links between the child node and the current node as next and prev respectively
            current_borrow.next = Some(child.clone());
            child.borrow_mut().prev = Some(current.clone());
        }

        // if no next node(at this point done processing the child node), next take from the stack(thus maintaining order)
        if current_borrow.next.is_none() && !stack.is_empty() {
            if let Some(next) = stack.pop() {
                current_borrow.next = Some(next.clone());
                next.borrow_mut().prev = Some(current.clone());
            }
        }

        // move to the next node
        runner = current_borrow.next.clone();
    }

    head
}


fn main(){
    let node1 = Node::new(1);
    let node2 = Node::new(2);
    let node3 = Node::new(3);
    let node4 = Node::new(4);
    let node5 = Node::new(5);
    let node6 = Node::new(6);
    let node7 = Node::new(7);
    let node8 = Node::new(8);
    let node9 = Node::new(9);
    let node10 = Node::new(10);
    let node11 = Node::new(11);
    let node12 = Node::new(12);


    if let Some(node1_ref) = node1.clone() {
        node1_ref.borrow_mut().next = node2.clone();
    }
    if let Some(node2_ref) = node2.clone() {
        node2_ref.borrow_mut().prev = node1.clone();
        node2_ref.borrow_mut().next = node3.clone();
    }
    if let Some(node3_ref) = node3.clone() {
        node3_ref.borrow_mut().prev = node2.clone();
        node3_ref.borrow_mut().next = node4.clone();
        node3_ref.borrow_mut().child = node7.clone();
    }
    if let Some(node4_ref) = node4.clone() {
        node4_ref.borrow_mut().prev = node3.clone();
        node4_ref.borrow_mut().next = node5.clone();
    }
    if let Some(node5_ref) = node5.clone() {
        node5_ref.borrow_mut().prev = node4.clone();
        node5_ref.borrow_mut().next = node6.clone();
    }
    if let Some(node6_ref) = node6.clone() {
        node6_ref.borrow_mut().prev = node5.clone();
    }
    if let Some(node7_ref) = node7.clone() {
        node7_ref.borrow_mut().next = node8.clone();
    }
    if let Some(node8_ref) = node8.clone() {
        node8_ref.borrow_mut().prev = node7.clone();
        node8_ref.borrow_mut().next = node9.clone();
        node8_ref.borrow_mut().child = node11.clone();
    }
    if let Some(node9_ref) = node9.clone() {
        node9_ref.borrow_mut().prev = node8.clone();
        node9_ref.borrow_mut().next = node10.clone();
    }
    if let Some(node10_ref) = node10.clone() {
        node10_ref.borrow_mut().prev = node9.clone();
    }
    if let Some(node11_ref) = node11.clone() {
        node11_ref.borrow_mut().next = node12.clone();
    }
    if let Some(node12_ref) = node12.clone() {
        node12_ref.borrow_mut().prev = node11.clone();
    }

    let flattened = flatten(node1);
    let mut runner = flattened;

    while let Some(current) = runner.clone(){
        print!("{} -> ", current.borrow().val);
        runner = current.borrow().next.clone();
    }
    println!("None")
}